bigtable/internal/testproxy/proxy.go

```diff
@@ -0,0 +1,976 @@
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+
+	rrq := req.GetRequest()
+	if rrq == nil {
+		return nil, stat.Error(codes.InvalidArgument, "request to CheckAndMutateRow() is missing inner request")
+	}
+
+	trueMuts := mutationFromProto(rrq.TrueMutations)
+	falseMuts := mutationFromProto(rrq.FalseMutations)
+
+	rfPb := rrq.PredicateFilter
+	f := bigtable.PassAllFilter()
+
+	if rfPb != nil {
+		f = *filterFromProto(rfPb)
+	}
+
+	c := bigtable.NewCondMutation(f, trueMuts, falseMuts)
+
+	t := btc.c.Open(rrq.TableName)
+	rowKey := string(rrq.RowKey)
+
+	ctx = btc.addCancelFunction(ctx)
+
+	var matched bool
+	ao := bigtable.GetCondMutationResult(&matched)
+
+	err := t.Apply(ctx, rowKey, c, ao)
+	if err != nil {
+		return nil, err
+	}
+	s.clientsLock.Unlock()
+
+	res := &pb.CheckAndMutateRowResult{
+		Status: &status.Status{
+			Code: int32(codes.OK),
+		},
+		Result: &btpb.CheckAndMutateRowResponse{
+			PredicateMatched: matched,
+		},
+	}
+	return res, nil
+}
+
+// SampleRowKeys responds to the SampleRowKeys RPC. This method gets a sampling
+// of the keys available in a table.
+func (s *goTestProxyServer) SampleRowKeys(ctx context.Context, req *pb.SampleRowKeysRequest) (*pb.SampleRowKeysResult, error) {
+	s.clientsLock.Lock()
+	btc, exists := s.clientIDs[req.ClientId]
+	if !exists {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+
+	rrq := req.GetRequest()
+	if rrq == nil {
+		return nil, stat.Error(codes.InvalidArgument, "request to CheckAndMutateRow() is missing inner request")
+	}
+
+	t := btc.c.Open(rrq.TableName)
+
+	ctx = btc.addCancelFunction(ctx)
+
+	keys, err := t.SampleRowKeys(ctx)
+	if err != nil {
+		return nil, err
+	}
+	s.clientsLock.Unlock()
+
+	sk := make([]*btpb.SampleRowKeysResponse, 0)
+	for _, k := range keys {
+		s := &btpb.SampleRowKeysResponse{
+			RowKey: []byte(k),
+		}
+		sk = append(sk, s)
+	}
+
+	res := &pb.SampleRowKeysResult{
+		Status: &status.Status{
+			Code: int32(codes.OK),
+		},
+		Sample: sk,
+	}
+	return res, nil
+}
+
+// ReadModifyWriteRow responds to the ReadModifyWriteRow RPC. This method
+// applies a non-idempotent change to a row.
+func (s *goTestProxyServer) ReadModifyWriteRow(ctx context.Context, req *pb.ReadModifyWriteRowRequest) (*pb.RowResult, error) {
+	s.clientsLock.Lock()
+	btc, exists := s.clientIDs[req.ClientId]
+	if !exists {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+
+	rrq := req.GetRequest()
+	if rrq == nil {
+		return nil, stat.Error(codes.InvalidArgument, "request to CheckAndMutateRow() is missing inner request")
+	}
+
+	t := btc.c.Open(rrq.TableName)
+	k := string(rrq.RowKey)
+	rpb := rrq.Rules
+
+	rmw := bigtable.NewReadModifyWrite()
+
+	for _, rp := range rpb {
+		switch r := rp.Rule; r.(type) {
+		case *btpb.ReadModifyWriteRule_AppendValue:
+			av := r.(*btpb.ReadModifyWriteRule_AppendValue)
```

Comment:
Don't need this for the reason noted above.

Reply:
The compiler gets unhappy if I remove the type cast.